runtime/major_gc.c

```diff
@@ -653,15 +658,61 @@ static intnat do_some_marking(intnat budget) {
   return budget;
 }
 
+static intnat pop_from_remark_stack(struct domain* trgt) {
+  /* handle remarking for a domain
+   * - if we own it, then we can remark it
+   * - if we don't own it and someone else does, give it to them
+   * - if it isn't owned by anyone, prevent it being used and then remark it
+   */
+  struct pool* p = caml_pop_pool_remark_for_domain(trgt);
+  if (p) {
+    struct domain* owner = caml_owner_of_pool(p);
+    if (owner == caml_domain_self()) {
+      caml_gc_log("Redarkening pool %p", p);
+      caml_redarken_pool(p, &mark_stack_push_act, 0);
+    } else if (owner != NULL) {
```

Comment:
I wanted a single `caml_pop_pool_remark_for_domain` function that worked for both local and global remark stacks. In the global remark stack, you can have pools which may not yet be orphaned (and so `owner != NULL`) but where the domain owning the pool is not accepting remark requests because it is in the process of orphaning its pools.

Reply:
Yep, missed that case. This was an attempt to see if we can get rid of some locking in the pop function, but then I realised we'd need to lock the remark stack anyway since other domains would be able to write to it.